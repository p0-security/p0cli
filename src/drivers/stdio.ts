/** Copyright Â© 2024-present P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/

/** Functions to handle stdio
 *
 * These are essentially wrappers around console.foo, but allow for
 * - Better testing
 * - Later redirection / duplication
 */

/** Used to output machine-readable text to stdout
 *
 * In general this should not be used for text meant to be consumed
 * only by humans.
 */
export function print1(message: any) {
  // eslint-disable-next-line no-console
  console.log(message);
}

/** Output human-consumable text to stderr
 *
 * In general this should not be used for machine-consumed text.
 */
export function print2(message: any) {
  // eslint-disable-next-line no-console
  console.error(message);
}
