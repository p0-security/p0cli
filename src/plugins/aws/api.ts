/** Copyright Â© 2024 P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/

export const AWS_API_VERSION = "2011-06-15";

export const arnPrefix = (account: string) => `arn:aws:iam::${account}`;
