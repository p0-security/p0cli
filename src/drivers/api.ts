/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Authn } from "../types/identity";
import { p0VersionInfo } from "../version";
import { getTenantConfig } from "./config";
import { defaultConfig } from "./env";
import { print2 } from "./stdio";
import * as path from "node:path";
import yargs from "yargs";

const DEFAULT_PERMISSION_REQUEST_TIMEOUT = 300e3; // 5 minutes

const tenantOrgUrl = (tenant: string) =>
  `${getTenantConfig()?.appUrl ?? defaultConfig.appUrl}/orgs/${tenant}`;
const tenantUrl = (tenant: string) => `${getTenantConfig().appUrl}/o/${tenant}`;
const publicKeysUrl = (tenant: string) =>
  `${tenantUrl(tenant)}/integrations/ssh/public-keys`;
const sshAuditUrl = (tenant: string) =>
  `${tenantUrl(tenant)}/integrations/ssh/audit`;

const commandUrl = (tenant: string) => `${tenantUrl(tenant)}/command/`;
const adminLsCommandUrl = (tenant: string) => `${tenantUrl(tenant)}/command/ls`;
export const tracesUrl = (tenant: string) => `${tenantUrl(tenant)}/traces`;

export const fetchOrgData = async <T>(orgId: string) =>
  unauthenticatedApiFetch<T>(tenantOrgUrl(orgId), "GET");

export const fetchAccountInfo = async <T>(authn: Authn) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/account`,
    method: "GET",
  });

export const fetchPermissionRequestDetails = async <T>(
  authn: Authn,
  requestId: string
) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/permission-requests/${requestId}`,
    method: "GET",
    maxTimeoutMs: DEFAULT_PERMISSION_REQUEST_TIMEOUT,
  });

export const fetchIntegrationConfig = async <T>(
  authn: Authn,
  integration: string
) =>
  baseFetch<T>(authn, {
    url: `${tenantUrl(authn.identity.org.slug)}/integrations/${integration}/config`,
    method: "GET",
  });

export const fetchStreamingCommand = async function* <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) {
  yield* fetchWithStreaming<T>(authn, {
    url: commandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
      wait: true,
    }),
  });
};

export const fetchCommand = async <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) =>
  baseFetch<T>(authn, {
    url: commandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
    }),
  });

/** Special admin 'ls' command that can retrieve results for all users. Requires 'owner' permission. */
export const fetchAdminLsCommand = async <T>(
  authn: Authn,
  args: yargs.ArgumentsCamelCase,
  argv: string[]
) =>
  baseFetch<T>(authn, {
    url: adminLsCommandUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      argv,
      scriptName: path.basename(args.$0),
    }),
  });

export const submitPublicKey = async <T>(
  authn: Authn,
  args: { publicKey: string; requestId: string }
) =>
  baseFetch<T>(authn, {
    url: publicKeysUrl(authn.identity.org.slug),
    method: "POST",
    body: JSON.stringify({
      requestId: args.requestId,
      publicKey: args.publicKey,
    }),
  });

export const fetchWithStreaming = async function* <T>(
  authn: Authn,
  args: {
    url: string;
    method: string;
    body?: string;
    maxTimeoutMs?: number;
  }
) {
  const token = await authn.getToken();
  const { version } = p0VersionInfo;
  const { url, method, body, maxTimeoutMs } = args;
  const fetchOptions = {
    method,
    headers: {
      authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": `P0 CLI/${version}`,
    },
    body,
    keepalive: true,
  };
  const textDecoder = new TextDecoder();
  try {
    const response = await fetch(
      url,
      maxTimeoutMs
        ? { ...fetchOptions, signal: AbortSignal.timeout(maxTimeoutMs) }
        : fetchOptions
    );
    const reader = response.body?.getReader();
    if (!reader) throw `No reader available`;
    while (true) {
      const read = await reader.read();
      if (read.done) {
        break;
      }
      const value = read.value;
      const text = textDecoder.decode(value);
      const jsonlSegments = text.split("\n").filter(Boolean);
      for (const segment of jsonlSegments) {
        const parsedResponse = JSON.parse(segment);
        if (parsedResponse.type === "error") {
          throw parsedResponse.error;
        }
        if (parsedResponse.type === "heartbeat") {
          continue;
        }
        if (parsedResponse.type !== "data" || !("data" in parsedResponse)) {
          throw "Invalid response from the server";
        }
        const { data } = parsedResponse;
        if ("error" in data) {
          throw data.error;
        }
        yield data as T;
      }
    }
  } catch (error) {
    if (
      error instanceof TypeError &&
      (error.message === "fetch failed" || error.message === "terminated")
    ) {
      throw `Network error: Unable to reach the server.`;
    } else {
      throw error;
    }
  }
};

export const auditSshSessionActivity = async (args: {
  authn: Authn;
  requestId: string;
  sshSessionId: string;
  action: `ssh.session.${"end" | "start"}`;
  debug: boolean | undefined;
}) => {
  const { authn, requestId, action, sshSessionId, debug } = args;

  if (debug) {
    print2(
      `Submitting audit log for request: ${requestId}, action: ${action}, sshSessionId: ${sshSessionId}`
    );
  }

  try {
    await baseFetch(authn, {
      url: sshAuditUrl(authn.identity.org.slug),
      method: "POST",
      body: JSON.stringify({
        requestId,
        action,
        sshSessionId,
      }),
    });
    if (debug) {
      print2(`Audit log submitted for request: ${requestId}`);
    }
  } catch (error) {
    if (debug) {
      print2(`Failed to submit audit log for request: ${requestId}`);
      print2(`Error: ${JSON.stringify(error)}`);
    }
  }
};

export const baseFetch = async <T>(
  authn: Authn,
  args: {
    url: string;
    method: string;
    body?: string;
    maxTimeoutMs?: number;
  }
) => {
  const { version } = p0VersionInfo;
  const { url, method, body, maxTimeoutMs } = args;
  const token = await authn.getToken();
  const fetchOptions = {
    method,
    headers: {
      authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      "User-Agent": `P0 CLI/${version}`,
    },
    body,
    keepalive: true,
  };

  try {
    const response = await fetch(
      url,
      maxTimeoutMs
        ? { ...fetchOptions, signal: AbortSignal.timeout(maxTimeoutMs) }
        : fetchOptions
    );
    const text = await response.text();
    const data = JSON.parse(text);
    if ("error" in data) {
      throw data.error;
    }
    return data as T;
  } catch (error) {
    if (error instanceof TypeError && error.message === "fetch failed") {
      throw `Network error: Unable to reach the server at ${url}.`;
    } else {
      throw error;
    }
  }
};

const unauthenticatedApiFetch = async <T>(
  url: string,
  method: string,
  body?: string
) => {
  const { version } = p0VersionInfo;
  try {
    const fetchConfig = {
      method,
      headers: {
        "Content-Type": "application/json",
        "User-Agent": `P0 CLI/${version}`,
      },
    };
    const response = await fetch(url, {
      ...fetchConfig,
      body,
    });
    const text = await response.text();
    const data = JSON.parse(text);
    if ("error" in data) {
      throw data.error;
    }
    return data as T;
  } catch (error) {
    if (error instanceof TypeError && error.message === "fetch failed") {
      throw `Network error: Unable to reach the server at ${url}.`;
    } else {
      throw error;
    }
  }
};
