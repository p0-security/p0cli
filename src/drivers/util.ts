import { print2 } from "./stdio";

/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
export const EXPIRED_CREDENTIALS_MESSAGE =
  "Your credentials have expired. Please run `p0 login <organization>` to refresh your credentials.";

// 0x0A is utf-8 character code of line feed
const LINE_FEED = 10;
/**
 * Converts a string that contains newline-delimited JSON
 * to an array of parsed json objects
 */
export const convertJsonlToArray = <T>(array: Uint8Array, maxErrors = 5) => {
  const out: T[] = [];
  const decoder = new TextDecoder();
  let offset = 0;
  let numErrors = 0;
  const totalLength = array.length;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const ix = array.indexOf(LINE_FEED, offset);
    if (ix < 0) {
      break;
    }
    const slice = array.slice(offset, ix);
    const json = decoder.decode(slice);
    try {
      if (json) out.push(JSON.parse(json));
    } catch (error) {
      numErrors += 1;
      print2("Failed to parse JSON line: " + json);
      if (numErrors >= maxErrors) {
        throw "Can not read streaming data";
      }
    }
    if (ix >= 0) {
      offset = ix + 1;
    } else {
      break;
    }
  }
  const remainingSegments =
    offset >= 0 && offset < totalLength
      ? array.slice(offset)
      : new Uint8Array();
  return {
    segments: out,
    remainingSegments,
  };
};
