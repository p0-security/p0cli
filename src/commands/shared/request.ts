/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { fetchCommand, fetchStreamingCommand } from "../../drivers/api";
import { authenticate } from "../../drivers/auth";
import { generateSpinUntil, print2, spinUntil } from "../../drivers/stdio";
import { Authn } from "../../types/identity";
import {
  PermissionRequest,
  PluginRequest,
  RequestResponse,
} from "../../types/request";
import { sys } from "typescript";
import yargs from "yargs";

export const PROVISIONING_ACCESS_MESSAGE =
  "Waiting for access to be provisioned";
export const EXISTING_ACCESS_MESSAGE = "Existing access found.";
export const ACCESS_EXISTS_ERROR_MESSAGE =
  "This principal already has this access";

const APPROVED = { message: "Your request was approved", code: 0 };
const DENIED = { message: "Your request was denied", code: 2 };
const ERRORED = { message: "Your request encountered an error", code: 1 };

const COMPLETED_REQUEST_STATUSES = {
  APPROVED,
  APPROVED_NOTIFIED: APPROVED,
  DONE: APPROVED,
  DONE_NOTIFIED: APPROVED,
  DENIED,
  ERRORED,
};

const isCompletedStatus = (
  status: any
): status is keyof typeof COMPLETED_REQUEST_STATUSES =>
  status in COMPLETED_REQUEST_STATUSES;

export const requestArgs = <T>(yargs: yargs.Argv<T>) =>
  yargs
    .parserConfiguration({ "unknown-options-as-args": true })
    .help(false) // Turn off help in order to forward the --help command to the backend so P0 can provide the available requestable resources
    .option("wait", {
      alias: "w",
      boolean: true,
      default: false,
      describe: "Block until the command is completed",
    })
    .option("arguments", {
      array: true,
      string: true,
      default: [] as string[],
    });

const resolveCode = (
  permission: PermissionRequest<PluginRequest>,
  logMessage: boolean
) => {
  const { status } = permission;
  if (isCompletedStatus(status)) {
    const { message, code } = COMPLETED_REQUEST_STATUSES[status];
    const errorMessage = permission.error
      ? `${message}: ${permission.error.message}`
      : message;
    if (code !== 0 || logMessage) print2(errorMessage);
    return code;
  } else {
    print2("Your request did not complete within 5 minutes.");
    return 4;
  }
};

export const request =
  (command: "grant" | "request") =>
  async <T>(
    args: yargs.ArgumentsCamelCase<{
      arguments: string[];
      wait?: boolean;
    }>,
    authn?: Authn,
    options?: {
      accessMessage?: string;
      message?: "all" | "approval-required" | "none" | "quiet";
    }
  ): Promise<RequestResponse<T> | undefined> => {
    const resolvedAuthn = authn ?? (await authenticate());
    const accessMessage = (message?: string) => {
      switch (message) {
        case "approval-required":
          return "Checking for access in P0";
        default:
          return "Requesting access";
      }
    };
    try {
      if (!args.wait) {
        const fetchCommandPromise = fetchCommand<RequestResponse<T>>(
          resolvedAuthn,
          args,
          [command, ...args.arguments]
        );
        const data =
          options?.message != "quiet"
            ? await spinUntil(
                accessMessage(options?.message),
                fetchCommandPromise
              )
            : await fetchCommandPromise;

        if (data && "ok" in data && "message" in data && data.ok) {
          const logMessage =
            !options?.message ||
            options?.message === "all" ||
            (options?.message === "approval-required" &&
              !data.isPreexisting &&
              !data.isPersistent);
          if (logMessage) print2(data.message);
          return data;
        } else {
          throw data;
        }
      } else {
        const fetchStreamingCommandPromise = fetchStreamingCommand<
          RequestResponse<T>
        >(resolvedAuthn, args, [command, ...args.arguments]);
        const fetchValue = async () => {
          const generatedValue = await fetchStreamingCommandPromise.next();
          if (generatedValue.done) {
            throw new Error("Could not get request id");
          }

          return generatedValue.value;
        };
        const data =
          options?.message != "quiet"
            ? await generateSpinUntil(
                accessMessage(options?.message),
                fetchStreamingCommandPromise
              )
            : await fetchValue();
        if (data && "ok" in data && "message" in data && data.ok) {
          const logMessage =
            !options?.message ||
            options?.message === "all" ||
            (options?.message === "approval-required" &&
              !data.isPreexisting &&
              !data.isPersistent);
          if (logMessage) print2(data.message);
          const finalData = await fetchValue();

          const code = resolveCode(
            finalData.request as PermissionRequest<PluginRequest>,
            logMessage
          );
          if (code) {
            sys.exit(code);
            return undefined;
          }
          return finalData;
        } else {
          throw data;
        }
      }
    } catch (error: any) {
      if (error instanceof Error && error.name === "TimeoutError") {
        print2("Your request did not complete within 5 minutes.");
        sys.exit(4);
      } else {
        print2(error);
        sys.exit(1);
      }
      return undefined;
    }
  };

export const provisionRequest = async (
  argv: yargs.ArgumentsCamelCase<{
    arguments: string[];
    wait?: boolean;
  }>,
  authn: Authn
) => {
  try {
    const response = await request("request")({ ...argv, wait: true }, authn, {
      message: "approval-required",
    });

    if (!response) {
      print2("Did not receive access ID from server");
      return;
    }

    const { id, isPreexisting } = response;

    print2(
      !isPreexisting ? PROVISIONING_ACCESS_MESSAGE : EXISTING_ACCESS_MESSAGE
    );
  } catch (error) {
    if (error === ACCESS_EXISTS_ERROR_MESSAGE) {
      print2(EXISTING_ACCESS_MESSAGE);
    } else {
      throw error;
    }
  }
};
