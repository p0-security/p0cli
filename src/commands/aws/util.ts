/** Copyright Â© 2025-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { print2 } from "../../drivers/stdio";
import { Authn } from "../../types/identity";
import { PluginRequest } from "../../types/request";
import { waitForProvisioning } from "../shared";
import { request } from "../shared/request";
import yargs from "yargs";

export const ACCESS_EXISTS_ERROR_MESSAGE =
  "This principal already has this access";

export const provisionRequest = async (
  argv: yargs.ArgumentsCamelCase<{
    arguments: string[];
    wait?: boolean;
  }>,
  authn: Authn
) => {
  const response = await request("request")(argv, authn, {
    message: "approval-required",
  });

  if (!response) {
    print2("Did not receive access ID from server");
    return;
  }

  const { id, isPreexisting } = response;

  print2(
    !isPreexisting
      ? "Waiting for access to be provisioned"
      : "Existing access found. Connecting to instance."
  );

  const provisionedRequest = await waitForProvisioning<PluginRequest>(
    authn,
    id
  );

  return { requestId: id, provisionedRequest };
};
