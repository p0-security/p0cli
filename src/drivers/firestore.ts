/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { Identity } from "../types/identity";
import { getContactMessage, loadConfig } from "./config";
import { bootstrapConfig } from "./env";
import { print2 } from "./stdio";
import { EXPIRED_CREDENTIALS_MESSAGE } from "./util";
import { FirebaseApp, FirebaseError, initializeApp } from "firebase/app";
import {
  getAuth,
  OAuthProvider,
  SignInMethod,
  signInWithCredential,
  UserCredential,
} from "firebase/auth";
import {
  collection as fsCollection,
  CollectionReference,
  doc as fsDoc,
  DocumentReference,
  getFirestore,
  terminate,
  Firestore,
} from "firebase/firestore";

const bootstrapApp = initializeApp(bootstrapConfig.fs, "bootstrapApp");
const bootstrapFirestore = getFirestore(bootstrapApp);

let app: FirebaseApp;
let firestore: Firestore;

export async function initializeFirebase() {
  if (!firestore) {
    const tenantConfig = await loadConfig();
    app = initializeApp(tenantConfig.fs, "authFirebase");
    firestore = getFirestore(app);
  }
}

const findProviderId = (identity: Identity) => {
  switch (identity.org.ssoProvider) {
    case "google":
      return SignInMethod.GOOGLE;
    case "google-oidc":
      return "oidc.google-oidc";
    default:
      return identity.org.providerId;
  }
};

export async function authenticateToFirebase(
  identity: Identity,
  options?: {
    debug?: boolean;
  }
): Promise<UserCredential> {
  const { credential } = identity;
  const tenantId = identity.org.tenantId;

  await initializeFirebase();

  const provider = new OAuthProvider(findProviderId(identity));

  const firebaseCredential = provider.credential({
    accessToken: credential.access_token,
    idToken: credential.id_token,
  });

  const auth = getAuth(app);
  auth.tenantId = tenantId;

  let userCredential;
  try {
    userCredential = await signInWithCredential(auth, firebaseCredential);
  } catch (error) {
    if (
      error instanceof FirebaseError &&
      error.code === "auth/invalid-credential"
    ) {
      throw EXPIRED_CREDENTIALS_MESSAGE;
    } else {
      if (options?.debug) {
        print2("Authentication error: " + error);
      }
      throw `An unexpected error occurred during authentication.\n${getContactMessage()}`;
    }
  }

  if (!userCredential?.user?.email) {
    throw `Can not sign in: this user has previously signed in with a different identity provider.\n${getContactMessage()}`;
  }

  return userCredential;
}

export const collection = <T>(path: string, ...pathSegments: string[]) => {
  return fsCollection(
    firestore,
    path,
    ...pathSegments
  ) as CollectionReference<T>;
};

export const doc = <T>(path: string) => {
  return fsDoc(firestore, path) as DocumentReference<T>;
};

export const bootstrapDoc = <T>(path: string) => {
  return fsDoc(bootstrapFirestore, path) as DocumentReference<T>;
};

/** Ensures that Firestore is shutdown at command termination
 *
 * This prevents Firestore from holding the command on execution completion or failure.
 */
export const fsShutdownGuard =
  <P, T>(cb: (args: P) => Promise<T>) =>
  async (args: P) => {
    try {
      await cb(args);
    } finally {
      shutdownFirebase();
    }
  };

export const shutdownFirebase = () => {
  if (bootstrapFirestore) void terminate(bootstrapFirestore);
  if (firestore) void terminate(firestore);
};
