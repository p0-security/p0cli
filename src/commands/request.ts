/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { fetchCommand } from "../drivers/api";
import { authenticate } from "../drivers/auth";
import { doc, guard } from "../drivers/firestore";
import { print2 } from "../drivers/stdio";
import { Authn } from "../types/identity";
import { Request, RequestResponse } from "../types/request";
import { onSnapshot } from "firebase/firestore";
import { sys } from "typescript";
import yargs from "yargs";

const WAIT_TIMEOUT = 300e3;

const APPROVED = { message: "Your request was approved", code: 0 };
const DENIED = { message: "Your request was denied", code: 2 };
const ERRORED = { message: "Your request encountered an error", code: 1 };

const COMPLETED_REQUEST_STATUSES = {
  APPROVED,
  APPROVED_NOTIFIED: APPROVED,
  DONE: APPROVED,
  DONE_NOTIFIED: APPROVED,
  DENIED,
  ERRORED,
};
const isCompletedStatus = (
  status: any
): status is keyof typeof COMPLETED_REQUEST_STATUSES =>
  status in COMPLETED_REQUEST_STATUSES;

const requestArgs = <T>(yargs: yargs.Argv<T>) =>
  yargs
    .parserConfiguration({ "unknown-options-as-args": true })
    .help(false) // Turn off help in order to forward the --help command to the backend so P0 can provide the available requestable resources
    .option("wait", {
      alias: "w",
      boolean: true,
      default: false,
      describe: "Block until the command is completed",
    })
    .option("arguments", {
      array: true,
      string: true,
      default: [] as string[],
    });

export const requestCommand = (yargs: yargs.Argv) =>
  yargs.command<{ arguments: string[] }>(
    "request [arguments..]",
    "Manually request permissions on a resource",
    requestArgs,
    guard(request)
  );

const waitForRequest = async (
  tenantId: string,
  requestId: string,
  logMessage: boolean
) =>
  await new Promise<number>((resolve) => {
    if (logMessage)
      print2("Will wait up to 5 minutes for this request to complete...");
    let cancel: NodeJS.Timeout | undefined = undefined;
    const unsubscribe = onSnapshot<Request, object>(
      doc(`o/${tenantId}/permission-requests/${requestId}`),
      (snap) => {
        const data = snap.data();
        if (!data) return;
        const { status } = data;
        if (isCompletedStatus(status)) {
          if (cancel) clearTimeout(cancel);
          unsubscribe?.();
          const { message, code } = COMPLETED_REQUEST_STATUSES[status];
          if (code !== 0 || logMessage) print2(message);
          resolve(code);
        }
      }
    );
    cancel = setTimeout(() => {
      unsubscribe?.();
      print2("Your request did not complete within 5 minutes.");
      resolve(4);
    }, WAIT_TIMEOUT);
  });

export const request = async (
  args: yargs.ArgumentsCamelCase<{
    arguments: string[];
    wait?: boolean;
  }>,
  authn?: Authn,
  options?: {
    message?: "all" | "approval-required" | "none";
  }
): Promise<RequestResponse | undefined> => {
  const resolvedAuthn = authn ?? (await authenticate());
  const { userCredential } = resolvedAuthn;
  const data = await fetchCommand<RequestResponse>(resolvedAuthn, args, [
    "request",
    ...args.arguments,
  ]);

  if (data && "ok" in data && "message" in data && data.ok) {
    const logMessage =
      !options?.message ||
      options?.message === "all" ||
      (options?.message === "approval-required" &&
        !data.isPreexisting &&
        !data.isPersistent);
    if (logMessage) print2(data.message);
    const { id } = data;
    if (args.wait && id && userCredential.user.tenantId) {
      const code = await waitForRequest(
        userCredential.user.tenantId,
        id,
        logMessage
      );
      if (code) {
        sys.exit(code);
        return undefined;
      }
      return data;
    } else return undefined;
  } else {
    throw data;
  }
};
