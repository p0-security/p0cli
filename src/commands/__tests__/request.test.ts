/** Copyright Â© 2024-present P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { fetchCommand } from "../../drivers/api";
import { print1, print2 } from "../../drivers/stdio";
import { failure } from "../../testing/yargs";
import { RequestResponse } from "../../types/request";
import { sleep } from "../../util";
import { requestCommand } from "../request";
import { onSnapshot } from "firebase/firestore";
import yargs from "yargs";

jest.mock("../../drivers/api");
jest.mock("../../drivers/auth");
jest.mock("../../drivers/stdio");

const mockFetchCommand = fetchCommand as jest.Mock;
const mockPrint1 = print1 as jest.Mock;
const mockPrint2 = print2 as jest.Mock;

describe("request", () => {
  beforeEach(() => jest.clearAllMocks());

  describe("when valid request command", () => {
    const command = "request gcloud role viewer";

    function mockFetch(response?: Partial<RequestResponse>) {
      mockFetchCommand.mockResolvedValue({
        ok: true,
        message: "a message",
        id: "abcefg",
        isPreexisting: false,
        isPersistent: false,
        ...(response ?? {}),
      });
    }

    describe.each([
      [false, false, true],
      [true, false, true],
      [false, true, true],
    ])(
      "preexisting=%p persistent=%p",
      (isPreexisting, isPersistent, should) => {
        it(`should${should ? "" : " not"} print request response`, async () => {
          mockFetch({ isPreexisting, isPersistent });
          await requestCommand(yargs()).parse(command);
          expect(mockPrint2.mock.calls).toMatchSnapshot();
          expect(mockPrint1).not.toHaveBeenCalled();
        });
      }
    );

    it("should wait for access", async () => {
      mockFetch();
      const promise = requestCommand(yargs()).parse(`${command} --wait`);
      const wait = sleep(10);
      await expect(wait).resolves.toBeUndefined();
      (onSnapshot as any).trigger({
        status: "DONE",
      });
      await expect(promise).resolves.toBeDefined();
      expect(mockPrint2.mock.calls).toMatchSnapshot();
      expect(mockPrint1).not.toHaveBeenCalled();
    });
  });

  describe("when error", () => {
    const command = "request foo";

    beforeAll(() => {
      mockFetchCommand.mockResolvedValue({
        error: `p0 request

Request access to a resource using P0

Commands:
  p0 request gcloud             Google Cloud

Options:
      --help    Show help                                              [boolean]
      --reason  Reason access is needed                                 [string]
  -w, --wait    Block until the request is completed                   [boolean]

Unknown argument: foo`,
      });
    });

    it("should print error message", async () => {
      const error = await failure(requestCommand(yargs()), command);
      expect(error).toMatchSnapshot();
    });
  });
});
