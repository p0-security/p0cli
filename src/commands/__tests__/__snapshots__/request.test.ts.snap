// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`request > when error > should handle stream errors 1`] = `[TypeError: terminated]`;

exports[`request > when error > should print error message 1`] = `
{
  "error": "p0 request

Request access to a resource using P0

Commands:
  p0 request gcloud             Google Cloud

Options:
      --help    Show help                                              [boolean]
      --reason  Reason access is needed                                 [string]
  -w, --wait    Block until the request is completed                   [boolean]

Unknown argument: foo",
}
`;

exports[`request > when valid request command > preexisting=%p persistent=%p false false > should print request response 1`] = `
[
  [
    "a message",
  ],
]
`;

exports[`request > when valid request command > preexisting=%p persistent=%p false true > should print request response 1`] = `
[
  [
    "a message",
  ],
]
`;

exports[`request > when valid request command > preexisting=%p persistent=%p true false > should print request response 1`] = `
[
  [
    "a message",
  ],
]
`;

exports[`request > when valid request command > should wait for access 1`] = `
[
  [
    "a message",
  ],
  [
    "Will wait up to 5 minutes for this request to complete...",
  ],
  [
    "Your request was approved",
  ],
]
`;
